### PR TITLE
Add new custom directory option to store copy of report on filesystem

### PR DESCRIPTION
**Pull request by tim-price**

### CODE REVIEW

File: edit_form.php [hunk 1]
```diff
@@ -98,13 +98,16 @@ public function definition() {
...
+        $mform->addElement('text', 'customdir', get_string('customdir', 'report_customsql'), 'size = 70');
         $mform->addElement('select', 'emailwhat', get_string('emailwhat', 'report_customsql'),
                 report_customsql_email_options());
         $mform->disabledIf('singlerow', 'runable', 'eq', 'manual');
         $mform->disabledIf('at', 'runable', 'ne', 'daily');
         $mform->disabledIf('emailto', 'runable', 'eq', 'manual');
+        $mform->disabledIf('customdir', 'runable', 'eq', 'manual');
         $mform->disabledIf('emailwhat', 'runable', 'eq', 'manual');
         $mform->setType('emailto', PARAM_RAW);
+        $mform->setType('customdir', PARAM_RAW);
```
Comment (timhunt) [inline edit_form.php]: Why is this not PARAM_PATH? Either there is a good reason, and we need a comment to explain why, and why the validation check is enough for security. Alternatively, it needs to be changed.

File: edit_form.php [hunk 2]
```diff
@@ -202,6 +205,13 @@ public function validation($data, $files) {
             $errors['querylimit'] = get_string('querylimitrange', 'report_customsql', REPORT_CUSTOMSQL_MAX_RECORDS);
         }
 
+        // Check that the custom directory is writable, if provided.
+        if (isset($data['customdir'])) {
+            if (!is_writable($data['customdir'])) {
```
Comment (timhunt) [inline edit_form.php]: Should we also have an is_dir here?

File: edit_form.php [hunk 3]
```diff
@@ -98,13 +98,16 @@ public function definition() {
                            get_string('onerow', 'report_customsql'));
 
         $mform->addElement('text', 'emailto', get_string('emailto', 'report_customsql'), 'size = 70');
+        $mform->addElement('text', 'customdir', get_string('customdir', 'report_customsql'), 'size = 70');
```
Comment (timhunt) [inline edit_form.php]: Should this field have a help icon, explaining that the file is saved in CSV format, and a bit about how the file name looks, etc?

File: locallib.php [hunk 4]
```diff
@@ -112,6 +112,16 @@ function report_customsql_generate_csv($report, $timenow) {
...
+    // Copy the file to the custom directory.
+    if (!empty($report->customdir)) {
+        list($csvfilename, $csvtimestamp) = report_customsql_csv_filename($report, $timenow);
+        $filename = $report->id . '-' . basename($csvfilename);
+        // Make sure we always have a working path.
+        $filepath = rtrim($report->customdir, '/');
+        $filepath = $filepath . '/' . $filename;
+        copy($csvfilename, $filepath);
+    }
+
```
Comment (timhunt) [inline locallib.php]: I think you should move this chunk to be next to the email bit, after the update record. That way, an error in copying the file does not stop the DB being updated.

Also, following the style around here, the chunk of cdoe inside the if statement should probably be extracted into a report_customsql_copy_csv_to_customdir. (You may be able to think of a better name.)

**Comment by timhunt** [comment]
Thanks for your work on this. That looks good. Just a few detailed comments, which hopefully should not take too long to sort out.

**Comment by timpricecatalyst** [comment]
I believe I have addressed all of the comments with the newest commit. If there's any further changes needed, let me know.

**Comment by timpricecatalyst** [comment]
I had not considered that case, as I thought it would always create a file. I have now added a check to see if the file exists before doing the copy, if it doesn't it just skips it.

**Comment by timhunt** [comment]
I'm sorry,it is probably getting to the point where all these comments are irritating, but I am afraid I still have more.

The problem is that between report_customsql_generate_csv and report_customsql_copy_csv_to_customdir you have built an unnecessary tangle.

You are calling report_customsql_copy_csv_to_customdir in both branches of the if ($csvfilenames) test, and then testing for whether the file exists in ..._copy_csv_.... However, that $csvfilenames test tells you whether any files exist. Therefore, all the logic could be much simpler.

Also, you really should move the if (!empty($report->customdir)) { test, so that is handled the same way as the if (!empty($report->emailto)) { test.

**Comment by timpricecatalyst** [comment]
I think I've fixed this now.

**Comment by brendanheywood** [comment]
hi @timhunt, I'm taking over this pull request from Tim Price. Is there anything still blocking outstanding to fix with this patch? I'm assuming you're fine with the idea of the patch conceptually?

**Comment by brendanheywood** [comment]
Actually feel free to kill this PR and I'll make a new one with a cleanly rebased single commit branch and fix up the version numbers which are now stale.

**Comment by timhunt** [comment]
That makes sense. Thanks. Closing.

**Comment by timhunt** [comment]
And, I have not looked at this code since the last comment above. I am afraid it will be that you make a commit that you think is right, then I will look at it and merge if possible.

**Pull request closed without merging**